cmd/proxy/main_test.go

```diff
@@ -0,0 +1,59 @@
+package main
+
+import (
+	"context"
+	"os"
+	"testing"
+
+	"github.com/go-kit/kit/log"
+)
+
+func TestHandleClient_Positive(t *testing.T) {
+	type subtest struct {
+		name    string
+		backend string
+
+		// One stream per query
+		streams []*MockQUICStream
+	}
+
+	subtests := []*subtest{
+		{
+			name:    "recursive_A",
+			backend: "8.8.4.4:53",
+			streams: []*MockQUICStream{
+				NewMockQUICStream(0, "ns1.com", "A", true, false),
+			},
+		},
+	}
+
+	for _, st := range subtests {
+		t.Run(st.name, func(tt *testing.T) {
+			ms := &MockQUICSession{
+				Streams: st.streams,
+			}
+
+			ctx := context.Background()
+			l := log.NewLogfmtLogger(log.NewSyncWriter(os.Stdout))
+			l = log.WithPrefix(l, "ts", log.DefaultTimestampUTC)
+
+			handleClient(l, ctx, ms, st.backend)
+
+			for _, stream := range ms.Streams {
+				if len(stream.OutBuf) == 0 {
+					t.Errorf("received empty response")
+				}
+
+				msg := stream.GetMsg()
+
+				if err := msg.Unpack(stream.OutBuf); err != nil {
```

Comment:
I find this very confusing: Some processing of the message already happens in `MockQUICStream` but this particular unpack operates on top of it's data members. All parsing should either be in the mocked implementation or the processing here should work with local variables without overwriting the context of `MockQuickStream`